src/deluge/model/clip/instrument_clip.cpp

```diff
@@ -2393,9 +2399,89 @@ void InstrumentClip::writeDataToFile(Serializer& writer, Song* song) {
 		writer.writeAttribute("mode", (char*)arpModeToString(arpSettings.mode));
 		writer.writeAttribute("syncLevel", arpSettings.syncLevel);
 		writer.writeAttribute("numOctaves", arpSettings.numOctaves);
-		writer.writeAttribute("spreadLock", arpSettings.spreadLock);
-		// Write locked spread params
+
+		if (output->type == OutputType::MIDI_OUT || output->type == OutputType::CV) {
+			writer.writeAttribute("gate", arpeggiatorGate);
+			writer.writeAttribute("rate", arpeggiatorRate);
+			// Community Firmware parameters (always write them after the official ones, just before closing the parent
+			// tag)
+			writer.writeAttribute("noteProbability", arpeggiatorNoteProbability);
+			writer.writeAttribute("bassProbability", arpeggiatorBassProbability);
+			writer.writeAttribute("chordProbability", arpeggiatorChordProbability);
+			writer.writeAttribute("ratchetProbability", arpeggiatorRatchetProbability);
+			writer.writeAttribute("ratchetAmount", arpeggiatorRatchetAmount);
+			writer.writeAttribute("sequenceLength", arpeggiatorSequenceLength);
+			writer.writeAttribute("chordPolyphony", arpeggiatorChordPolyphony);
+			writer.writeAttribute("rhythm", arpeggiatorRhythm);
+			writer.writeAttribute("spreadVelocity", arpeggiatorSpreadVelocity);
+			writer.writeAttribute("spreadGate", arpeggiatorSpreadGate);
+			writer.writeAttribute("spreadOctave", arpeggiatorSpreadOctave);
+		}
+
+		// Community Firmware parameters (always write them after the official ones, just before closing the parent tag)
+		writer.writeAttribute("syncType", arpSettings.syncType);
+		writer.writeAttribute("arpMode", (char*)arpModeToString(arpSettings.mode));
+		writer.writeAttribute("chordType", arpSettings.chordTypeIndex);
+		writer.writeAttribute("noteMode", (char*)arpNoteModeToString(arpSettings.noteMode));
+		writer.writeAttribute("octaveMode", (char*)arpOctaveModeToString(arpSettings.octaveMode));
+		writer.writeAttribute("mpeVelocity", (char*)arpMpeModSourceToString(arpSettings.mpeVelocity));
+		writer.writeAttribute("stemRepeat", arpSettings.numStepRepeats);
+		writer.writeAttribute("randomizerLock", arpSettings.randomizerLock);
+
+		// Write locked params
 		char buffer[9];
+		// Note probability
+		writer.insertCommaIfNeeded();
+		writer.write("\n");
+		writer.printIndents();
+		writer.writeTagNameAndSeperator("lockedNoteProb");
+		writer.write("\"0x");
```

Comment:
Can you use `writeAttributeHexBytes` here instead of the more custom serialization code?

It might require splitting `lockedNoteProb` (and `locked{Base,Chord,Ratchet}Prob`) into 2 separate entries but I think that's likely a good idea anyway and it will make evolving the serialization code less painful if the higher level functions are in use.